src/Npgsql/PoolManager.cs

```diff
@@ -71,16 +72,36 @@ internal static ConnectorPool GetOrAdd(string key, ConnectorPool pool)
                     return result;
 
                 // May need to grow the array.
-                if (_nextSlot == _pools.Length)
+                if (_nextSlot == _poolsAliases.Length)
                 {
-                    var newPools = new (string, ConnectorPool)[_pools.Length * 2];
+                    var newPools = new (string, ConnectorPool)[_poolsAliases.Length * 2];
+                    Array.Copy(_poolsAliases, newPools, _poolsAliases.Length);
+                    _poolsAliases = newPools;
+                }
+
+                _poolsAliases[_nextSlot].Key = key;
+                _poolsAliases[_nextSlot].Pool = pool;
+                Interlocked.Increment(ref _nextSlot);
+
+                var idx = 0;
+                ConnectorPool? existsPool = null;
+                for (; idx < _pools.Length; idx++)
```

Comment:
Note that in NpgsqlConnection we've already found the pool via the canonical/normalized connection string, so we could just call a different method here rather than searching for it again.

Reply:
At this point, we check if we saved the new pool in _pools array earlier and if not, we look for a free slot in _pools array for it.

Reply:
Not sure if I understood correctly ... Made an additional parameter to let me know if a new pool is being registered or an existing one.